### ISSUE TITLE
Actor Culling doesn't take animations into account

### ISSUE THREAD

**Issue by drewc5131** [panda3d/panda3d#1083]
## Description
When setFinal = 0 on an actor, culling doesn't take into account the position of the model parts when moved in an animation.

## Steps to Reproduce
[AvatarCulling.zip](https://github.com/panda3d/panda3d/files/5742438/AvatarCulling.zip)
Set setFinal to False, and move the actor to the side of the screen. In this example, the T-Pose would show the wrist off screen, so it stops rendering even if it is actually onscreen.

https://user-images.githubusercontent.com/31282527/103142524-86963700-46d2-11eb-9d07-306f24032420.mp4
This becomes an even bigger problem on some other models, where the entire hand will disappear

**Comment by rdb** [comment]
This is a tricky problem, since the actor pose is calculated by the cull process, but the cull process depends on the actor pose.  An animation could theoretically move a part of the actor offscreen from one frame to the next, so this would mean that we can't really effectively apply hierarchical culling to actors.

We could calculate a bounding volume that encompasses all poses the actor could possibly be in, but even that's not complete because it's possible to use controlJoint and move a particular body part far into the distance.

The workaround, until we think of a good solution, is to manually apply a bounding volume that you know encompasses the actor in all its potential animations.  This would still be my recommendation even after we fix this bug properly, because you would be able to prevent Panda from calculating the pose when the actor is out of view, which is otherwise not possible.

**Comment by paustian** [comment]
This sounds interesting. Want me to take a look? I am a big player of toontown so would be happy to help out with any issues they are having.

**Comment by rdb** [comment]
Feel free to take a look, though this is a gnarly issue and would like to discuss any ideas for tackling this before starting implementation.

**Comment by paustian** [comment]
I will take a look. First note is that I find it interesing that the hands are drawn, but not the wrist. I will locate the cull process and start digging through it. Is this in the python code or C++?

**Comment by rdb** [comment]
Consolidating the notes I sent in Discord, for the record:

The cull system is implemented in C++, the relevant parts being [PandaNode::update_cached()](https://github.com/panda3d/panda3d/blob/089a615adb401f1b91fd0fc7138aa65f2abd769b/panda/src/pgraph/pandaNode.cxx#L3254) for bounding volume propagation, [Character::cull_callback()](https://github.com/panda3d/panda3d/blob/089a615adb401f1b91fd0fc7138aa65f2abd769b/panda/src/char/character.cxx#L200) to trigger the character update and the most important bits of the cull process itself are the CullTraverserData and CullTraverser classes, both in panda/src/pgraph.

The problem is one of design, a combination of the factors:
1. Bounding volumes propagate upwards through the scene graph, ie. if you move a child node then the bounding volumes all the way to the root are recalculated, by PandaNode::update_cached().
2. The cull process doesn't visit child nodes that it knows are out of view (an important optimisation)
3. The animation is being updated by the cull callback, which is called when the cull traverser visits the character node.

So if the character is out of view, the cull process won't visit it, and the animation won't update, which means that Panda won't know if the animation would have caused parts of the model to fly into the camera's view.

If the Character were to somehow tell the cull process "always visit this node", that would fix the problem.  But the cost to performance could be quite significant: even if you're looking away from a crowd, Panda would still be calculating all of the crowd's animations.

So, at this point I think we need some ideas for a better solution.

**Comment by paustian** [comment]
Just Dropping a test case implemented in C++ that shows the culling error. This will allow stepping through the code.
[AvatarTestor.cpp.zip](https://github.com/panda3d/panda3d/files/6795717/AvatarTestor.cpp.zip)

**Comment by paustian** [comment]
I played around with this a bit today and have some questions. First, let me rephrase what you said so that I am hopefully understanding what's going on. All of the drawing process starts at 

void GraphicsEngine::WindowRenderer::do_frame

It culls, draws, and then does events. During the cull process, each model is checked to see if it needs to be animated, if so, it is then animated. The problem comes in that the cull process is doing the cull and then animating the model. Culling is done first, to save processing power. If a model is offscreen, we don't want to animate it, but animation may bring it back into view!

As you said, one brute force solution to allow characters to request to always be animated. This is unacceptable because it will cause a huge performance hit. However, what if the visibility from the previous cull process was stored in a new class variable on the node (_was_visible) and only animate those, _before_ we do the cull processing. That would put the character (or any other node) in the right pose before culling was calculated. It would also only apply to things we knew were visible and would not be a huge performance hit.

I am also a little confused about what is going on with the example case. I thought I understood what is going on here. The cull process is calculating the bounding volume based upon the T pose. When you get close to the edge, the sleeve would thus be out of frame in the T-pose and the culling process determines it should not be drawn. But the call to set pose puts it into the walk pose and now it is in frame and should draw it. But I have some questions...

1. Why doesn't the culling process readjust to the walk animation bounding box? Why does it continue to use the T-pose bounding box? Are bounding volumes not updated as the animation changes? Is this another performance optimization?
2. Why do you only lose the sleeve? Shouldn't the hand also be lost? 
3. Is there an easy way to know as you step through the code, what part of the model is being culled? (Hand, sleeve, suit, etc) If you can point me to a reference that explains this, or explain it yourself that would help. Thanks for any advice.

**Comment by paustian** [comment]
I added show_bounds to the cogActor and now I see what's going on. The bounding volume is a big sphere that encompasses all possible positions of the actor. I guess if part of the bounding volume is within the frame, it is queued for possible culling.

**Comment by Moguri** [comment]
It seems to me that  two issues are being discussed here:
1. For performance reasons (especially with hardware skinning) it is not feasible to recalculate the bounds of an animated mesh every frame, which means that an Actor can get culled even when a part of it is still in view.
2. When an Actor contains multiple meshes, due to similar reasons as 1, some of the meshes may get culled in an undesirable way.

I believe these are ultimately the same underlying cause (transformed verts are out of sync with their bounding volumes), but might have different solutions. I do not believe we will ever be able to tackle the root cause (transformed verts being out of sync with culling bounds) without serious performance issues, but we might be able to get a "good enough" solution by tackling some of the  symptoms.

For 1, we might be able to get away with adding a fudge factor when computing bounds for actors/characters. This could be a user configurable value to scale the bounds in the hopes that we get more pose variations within the expanded bounds. If an animation contains a large/long stretch, then this wouldn't help, but it should account for differences between say a rest and walking post. With bounds not as tight, the character may not cull as soon as it maybe should, but will still get culled eventually, so I think we would be okay.

For 2, we might need to look into something to prevent culling if other parts of the character are in view
